homeassistant/components/konnected.py

```diff
@@ -0,0 +1,312 @@
+from homeassistant.helpers import discovery
+from homeassistant.helpers import config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+REQUIREMENTS = ['konnected==0.1.2']
+
+DOMAIN = 'konnected'
+
+PIN_TO_ZONE = {1: 1, 2: 2, 5: 3, 6: 4, 7: 5, 8: 'out', 9: 6}
+ZONE_TO_PIN = {zone: pin for pin, zone in PIN_TO_ZONE.items()}
+
+_BINARY_SENSOR_SCHEMA = vol.All(
+    vol.Schema({
+        vol.Exclusive(CONF_PIN, 's_pin'): vol.Any(*PIN_TO_ZONE),
+        vol.Exclusive(CONF_ZONE, 's_pin'): vol.Any(*ZONE_TO_PIN),
+        vol.Required(CONF_TYPE): DEVICE_CLASSES_SCHEMA,
+        vol.Optional(CONF_NAME): cv.string,
+    }), cv.has_at_least_one_key(CONF_PIN, CONF_ZONE)
+)
+
+_SWITCH_SCHEMA = vol.All(
+    vol.Schema({
+        vol.Exclusive(CONF_PIN, 'a_pin'): vol.Any(*PIN_TO_ZONE),
+        vol.Exclusive(CONF_ZONE, 'a_pin'): vol.Any(*ZONE_TO_PIN),
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Optional('activation', default='high'):
+            vol.All(vol.Lower, vol.Any('high', 'low'))
+    }), cv.has_at_least_one_key(CONF_PIN, CONF_ZONE)
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema({
+            vol.Required('auth_token'): cv.string,
+            vol.Required(CONF_DEVICES): [{
+                vol.Required(CONF_ID): cv.string,
+                vol.Optional(CONF_BINARY_SENSORS): vol.All(
+                    cv.ensure_list, [_BINARY_SENSOR_SCHEMA]),
+                vol.Optional(CONF_SWITCHES): vol.All(
+                    cv.ensure_list, [_SWITCH_SCHEMA]),
+            }],
+        }),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+DEPENDENCIES = ['http', 'discovery']
+
+ENDPOINT_ROOT = '/api/konnected'
+UPDATE_ENDPOINT = (ENDPOINT_ROOT + r'/device/{device_id:[a-zA-Z0-9]+}')
+
+
+async def async_setup(hass, config):
+    """Set up the Konnected platform."""
+    cfg = config.get(DOMAIN)
+    if cfg is None:
+        cfg = {}
+
+    auth_token = cfg.get('auth_token')
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {'auth_token': auth_token}
+
+    async def async_device_discovered(service, info):
+        """Call when a Konnected device has been discovered."""
+        _LOGGER.debug("Discovered a new Konnected device: %s", info)
+        host = info.get(CONF_HOST)
+        port = info.get(CONF_PORT)
+
+        device = KonnectedDevice(hass, host, port, cfg)
+        await device.async_setup()
+
+    discovery.async_listen(
+        hass,
+        SERVICE_KONNECTED,
+        async_device_discovered)
+
+    hass.http.register_view(KonnectedView(auth_token))
+
+    return True
+
+
+class KonnectedDevice(object):
+    """A representation of a single Konnected device."""
+
+    def __init__(self, hass, host, port, config):
+        """Initialize the Konnected device."""
+        self.hass = hass
+        self.host = host
+        self.port = port
+        self.user_config = config
+
+        import konnected
+        self.client = konnected.Client(host, str(port))
```

Comment:
What happens if the device is unavailable?

Reply:
The device depends on the `discovery` component to be recognized and initialized when HASS starts up, so if the device is unavailable at startup, it would never reach this code.

It is possible that the device is online at HASS startup but then later goes offline -- in this case it obviously would no longer push state updates to HASS, and signals from HASS to trigger switches would timeout. I think this is not a big problem for now. In the future I'm thinking of adding some retry logic and possibly device health/status with notification if a device goes offline. I think that's too much (potentially unnecessary) complexity for this first release though.